lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,181 @@
+package graphql.nadel.validation
+
+import graphql.Scalars
+import graphql.nadel.dsl.RemoteArgumentDefinition
+import graphql.nadel.dsl.UnderlyingServiceHydration
+import graphql.nadel.engine.util.isList
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.nadel.engine.util.unwrapOne
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
+        /*
+*  Checks the type of a hydration argument derived from either a $source field output type or $argument field input
+* argument type against the type of an actor field argument to see if they match
+* will also allow for cases of batched hydration eg ID! -> [ID]
+*/
+        fun getHydrationInputErrors(
```

Comment:
Have you considered batch hydration here? It's possible to have `users(ids: [ID!]!): [User!]` be the actor field but have `$source.userId` that references a `userId: ID` i.e. `[ID!]!` vs `ID` but it is still compatible.

For basic hydration this looks good.

Nit: please run the IntelliJ formatter on the code :)

Reply:
can you elaborate on that example above so I can make a test case?
Currently if we have the required input as `[ID!]!` but supply just an `ID` then it will fail validation. Why are they supposed to be compatible in the case of batch hydration?

I don't see how batch vs non-batch makes a different to the test since we are just checking the input types

Reply:
added the two missing cases as discussed offline:

```
//could have ID feed into [ID] (as a possible batch hydration case).
            // in this case we need to unwrap the list and check types
            if (isBatchHydration && (!unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList)) {
```
and
```
//could have [ID] feed into ID (non-batch, ManyToOne case explained in NadelHydrationStrategy.kt)
            // in this case we need to unwrap the list and check types
            else if (!isBatchHydration && (unwrappedHydrationSourceFieldType.isList && !unwrappedActorFieldArgType.isList)) {
```